internal/v4/auth.go

```diff
@@ -75,6 +76,43 @@ func (h *ReqHandler) authorize(req *http.Request, acl []string, alwaysAuth bool,
 	return authorization{}, httpbakery.NewDischargeRequiredErrorForRequest(m, cookiePath, verr, req)
 }
 
+func (h *ReqHandler) checkTerms(id *router.ResolvedURL, req *http.Request) error {
+	user, passwd, err := parseCredentials(req)
+	if err == nil {
+		if user != h.handler.config.AuthUsername || passwd != h.handler.config.AuthPassword {
+			return errgo.WithCausef(nil, params.ErrUnauthorized, "invalid user name or password")
+		}
+		return nil
+	}
+	entity, err := h.Store.FindEntity(id)
+	if err != nil {
+		return errgo.Mask(err, errgo.Is(params.ErrNotFound))
+	}
+	if len(entity.CharmMeta.Terms) > 0 {
+		bk := h.Store.Bakery
+		if h.handler.config.TermsLocation == "" || bk == nil {
+			return errgo.New("charmstore not serving charms requiring agreement to terms and conditions")
+		}
+
+		_, verr := httpbakery.CheckRequest(bk, req, nil, checkers.OperationChecker("get-archive"))
+		if verr == nil {
+			return nil
+		}
+		if _, ok := errgo.Cause(verr).(*bakery.VerificationError); !ok {
+			return errgo.Mask(verr)
+		}
+		m, err := bk.NewMacaroon("", nil, []checkers.Caveat{
```

Comment:
It seems a bit of a pity that we'll be generating two macaroons that both guard access to the same entity when theoretically one would do. Apart from the fact that it implies more round trips, this increases the likelihood of running out of client cookiejar space.

I'm also not that keen that we're calling CheckRequest twice and substantially duplicating the logic in checkRequest. 

I think that perhaps we should add support for authorising specified operations into AuthorizeEntity, so there's only one place that generates macaroons. This might require a bit of thought.